lib/IRGen/GenCall.cpp

```diff
@@ -373,6 +373,70 @@ static void addIndirectResultAttributes(IRGenModule &IGM,
   attrs = attrs.addParamAttributes(IGM.getLLVMContext(), paramIndex, b);
 }
 
+CombinedResultAndErrorType irgen::combineResultAndTypedErrorType(
+    const IRGenModule &IGM, const NativeConventionSchema &resultSchema,
+    const NativeConventionSchema &errorSchema) {
+  CombinedResultAndErrorType result;
+  SmallVector<llvm::Type *, 8> elts;
+  resultSchema.enumerateComponents(
+      [&](clang::CharUnits offset, clang::CharUnits end, llvm::Type *type) {
+        elts.push_back(type);
+      });
+
+  SmallVector<llvm::Type *, 8> errorElts;
+  errorSchema.enumerateComponents(
+      [&](clang::CharUnits offset, clang::CharUnits end, llvm::Type *type) {
+        errorElts.push_back(type);
+      });
+
+  llvm::SmallVector<llvm::Type *, 4> combined;
+
+  auto resIt = elts.begin();
+  auto errorIt = errorElts.begin();
+
+  while (resIt < elts.end() && errorIt < errorElts.end()) {
+    auto *res = *resIt;
+    if (!res->isIntOrPtrTy()) {
+      combined.push_back(res);
+      ++resIt;
+      continue;
+    }
+
+    auto *error = *errorIt;
```

Comment:
Can you add a comment somewhere what the preconditions are for this function?

I got sidetrack a little bit reading this code the first time because I assumed that we could have an `error` type that is a `float` which would be problematic for this code.

Then I realized we only get here if `errorSchema.shouldReturnTypedErrorIndirectly()` returns false (I assume) which means that all possible error types are either Int or Ptr.

I think you should also add an assert that an `error` type never is anything other those two.